### PR TITLE
[webgpu] Enable timestamp query API for profiling

### PR DESCRIPTION
**Pull request by haoyunfeix**
FEATURE

Use timestamp query API in webgpu backend for profiling

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4383)

### CODE REVIEW

**Comment by haoyunfeix** [comment]
@qjia7 @annxingyuan @tafsiri @pyu10055 PTAL

**Comment by qjia7** [comment]
LGTM.  @kainino0x take a look. Thanks.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 1]
```diff
@@ -453,11 +463,26 @@ export class WebGPUBackend extends KernelBackend {
...
+    }
     pass.setPipeline(pipeline);
     pass.setBindGroup(0, bg);
     pass.dispatch(
         program.dispatch[0], program.dispatch[1], program.dispatch[2]);
+    if (shouldTimeProgram) {
+      if (this.supportTimeQuery) {
+        pass.writeTimestamp(this.querySet, 1);
+      } else {
+        query = this.endTimer(query);
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:481]: If I'm reading this right, it doesn't make sense to endTimer here; nothing has actually happened between startTimer and endTimer except a few encoding commands.

Even the old code doesn't make much sense as it contains the submit, but it doesn't wait for anything to happen.
I think you should consider removing the CPU timing entirely. Or, keeping it as a separate timing stat that's always enabled - not an alternative to GPU timing.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 2]
```diff
@@ -479,12 +504,40 @@ export class WebGPUBackend extends KernelBackend {
...
   }
+  async getTimeFromQuerySet(querySet: GPUQuerySet) {
+    const queryBuffer = this.acquireBuffer(16,
+            GPUBufferUsage.COPY_SRC | GPUBufferUsage.QUERY_RESOLVE);
+    const dst = this.acquireBuffer(
+        16, GPUBufferUsage.MAP_READ | GPUBufferUsage.COPY_DST);
+
+    const encoder = this.device.createCommandEncoder();
+    // tslint:disable-next-line:no-any
+    (encoder as any).resolveQuerySet(querySet, 0, 2, queryBuffer, 0);
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:526]: Sorry this update hasn't made it into the types yet.

**Review by kainino0x — APPROVED** [review]

**Pull request merged**